runtime/tests/checker/for_test.go

```diff
@@ -68,7 +68,9 @@ func TestCheckForEmpty(t *testing.T) {
       }
     `)
 
-	assert.NoError(t, err)
+	assert.Error(t, err)
+	errs := ExpectCheckerErrors(t, err, 1)
+	assert.IsType(t, &sema.TypeAnnotationRequiredError{}, errs[0])
```

Comment:
Why is this producing an error now? If the problem is that there is no contextual information: We could expect this to be `[AnyStruct]`, as iteration in this construct is currently always over arrays and iteration over resources is not supported.

Then this change here and in https://github.com/onflow/cadence/pull/1027/files#diff-ec994e4f2ba53bd43232ba59cf3f9f4c83f36bed61c097893e6d2b5c89eb25b3L202 could be undone.

Reply:
Yeah, the catch is we can't always pass `[AnyStruct]` as the expected type, because then it would cause other errors (i.e: imprecise inferring for arrays, etc. fixed-length arrays, etc). So we need to specially check for an empty array, and then pass `[AnyStruct]` as the expected type. Didn't really like special-casing there, but maybe it makes sense to do so 👍

Reply:
Updated in a961aae0e157156a17984b134d434240e84e67eb

Reply:
Ah, I see. Yeah, special casing is a bit unfortunate, but I think the cost is OK for keeping backwards compatibility.